tensorhive/core/services/ProtectionService.py

```diff
@@ -70,39 +103,65 @@ def as_dict(line):
 
         return [as_dict(line) for line in stdout_lines]
 
+    def find_hostname(self, uuid: str) -> str:
+        '''Seeks the hostname of node with GPU given by UUID'''
+        infrastructure = self.infrastructure_manager.infrastructure
+        for hostname, data in infrastructure.items():
+            if data.get('GPU') and data['GPU'].get(uuid):
+                return hostname
+        log.info('GPU with UUID="{}" was not found'.format(uuid))
+        return None
+
+    @property
+    def ignored_processes(self):
+        return [
+            '/usr/lib/xorg/Xorg',
+            'X'
+        ]
+
+    def process_owners_on_node(self, node_processes) -> List[str]:
+        result = []
+        for uuid, processes in node_processes.items():
+            for process in processes:
+                if process['command'] not in self.ignored_processes:
+                    result.append(process['owner'])
+        return result
+
     @override
     def do_run(self):
         time_func = time.perf_counter
         start_time = time_func()
 
         # 1. Get list of current reservations
-        #current_reservations = ReservationEventModel.current_events()
+        current_reservations = ReservationEventModel.current_events()
 
-        # Mock (it only imitates result from database, it won't be a dict!)
-        current_reservations = [
-            {
-                'node': {'hostname': 'localhost'},
-                'user': {'username': 'UNPERMITTED_USERNAME_MOCK'}
-            }
-        ]
+        # DEBUG ONLY
+        __reservations_as_dict = [r.as_dict for r in current_reservations]
+        import json
+        log.debug(json.dumps(__reservations_as_dict, indent=4))
 
         unauthorized_sessions = []
         for reservation in current_reservations:
             # 1. Extract reservation info
-            hostname = reservation['node']['hostname']
-            username = reservation['user']['username']
+            hostname = self.find_hostname(uuid=reservation.resource_id)
+            username = UserModel.find_by_id(reservation.user_id).username
+            if hostname is None or username is None:
+                log.warning('Unable to process the reservation ({}@{}), skipping...'.format(username, hostname))
+                continue
 
             # 2. Establish connection to node and find all tty sessions
             node_connection = self.connection_manager.single_connection(hostname)
             node_sessions = self.node_tty_sessions(node_connection)
+            node_processes = self.node_gpu_processes(hostname)
+            process_owners = self.process_owners_on_node(node_processes)
 
             # 3. Any session that does not belong to a priviliged user should be rembered
             for session in node_sessions:
-                if session['USER'] != username:
+                if (session['USER'] != username) and (session['USER'] in process_owners):
```

Comment:
Is the specific id for the reserved device matched with the id of the device used by the process?

Reply:
Fixed, now it should be working as expected.